core/finality-grandpa/src/light_import.rs

```diff
@@ -0,0 +1,736 @@
+	// => justification is verified using current authorities set + we could proceed further
+	//
+	// pessimistic scenario: the GRANDPA authorities set has changed
+	// => we need to fetch new authorities set (i.e. finality proof) from remote node
+
+	// first, try to behave optimistically
+	let authority_set_id = data.authority_set.set_id();
+	let justification = J::decode_and_verify(
+		&justification,
+		authority_set_id,
+		&data.authority_set.authorities(),
+	);
+
+	// BadJustification error means that justification has been successfully decoded, but
+	// it isn't valid within current authority set
+	let justification = match justification {
+		Err(ClientError::BadJustification(_)) => {
+			trace!(
+				target: "finality",
+				"Justification for {} is not valid within current authorities set. Requesting finality proof.",
+				hash,
+			);
+
+			let mut imported_aux = ImportedAux::default();
+			imported_aux.needs_finality_proof = true;
+			return Ok(ImportResult::Imported(imported_aux));
+		},
+		Err(e) => {
+			trace!(
+				target: "finality",
+				"Justification for {} is not valid. Bailing.",
+				hash,
+			);
+
+			return Err(ConsensusErrorKind::ClientImport(e.to_string()).into());
+		},
+		Ok(justification) => {
+			trace!(
+				target: "finality",
+				"Justification for {} is valid. Finalizing the block.",
+				hash,
+			);
+
+			justification
+		},
+	};
+
+	// finalize the block
+	do_finalize_block(client, data, hash, number, justification.encode())
+}
+
+/// Finalize the block.
+fn do_finalize_block<B, E, Block: BlockT<Hash=H256>, RA>(
+	client: &Client<B, E, Block, RA>,
+	data: &mut LightImportData<Block>,
+	hash: Block::Hash,
+	number: NumberFor<Block>,
+	justification: Justification,
+) -> Result<ImportResult, ConsensusError>
+	where
+		B: Backend<Block, Blake2Hasher> + 'static,
+		E: CallExecutor<Block, Blake2Hasher> + 'static + Clone + Send + Sync,
+		RA: Send + Sync,
+		NumberFor<Block>: grandpa::BlockNumberOps,
+{
+	// finalize the block
+	client.finalize_block(BlockId::Hash(hash), Some(justification), true).map_err(|e| {
+		warn!(target: "finality", "Error applying finality to block {:?}: {:?}", (hash, number), e);
+		ConsensusError::from(ConsensusErrorKind::ClientImport(e.to_string()))
+	})?;
+
+	// forget obsoleted consensus changes
+	let consensus_finalization_res = data.consensus_changes
+		.finalize((number, hash), |at_height| canonical_at_height(&client, (hash, number), true, at_height));
+	match consensus_finalization_res {
+		Ok((true, _)) => require_insert_aux(
+			&client,
+			LIGHT_CONSENSUS_CHANGES_KEY,
+			&data.consensus_changes,
+			"consensus changes",
+		)?,
+		Ok(_) => (),
+		Err(error) => return Err(on_post_finalization_error(error, "consensus changes")),
+	}
+
+	// update last finalized block reference
+	data.last_finalized = hash;
+
+	Ok(ImportResult::imported())
+}
+
+/// Load light impoty aux data from the store.
+fn load_aux_import_data<B, Block: BlockT<Hash=H256>, PRA>(
```

Comment:
I thought that `aux_schema` is already bloated a bit, now as it supports version migrations. I.e it will be bigger than `light_import.rs` if we'll move this. I could move to new `aux_schema_light.rs`, if you insist, though.